onnxruntime/core/providers/cuda/math/matmul_integer.cc

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "matmul_integer.h"
+#include "matmul_integer.cuh"
+#include "core/providers/cpu/math/matmul_helper.h"
+#include "core/providers/cuda/shared_inc/fpgeneric.h"
+#include "core/providers/cuda/cuda_allocator.h"
+#include "core/providers/common.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_TYPED_KERNEL_EX(
+    MatMulInteger,
+    kOnnxDomain,
+    10,
+    int8_t,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .TypeConstraint("T1", DataTypeImpl::GetTensorType<int8_t>())
+        .TypeConstraint("T2", DataTypeImpl::GetTensorType<int8_t>())
+        .TypeConstraint("T3", DataTypeImpl::GetTensorType<int32_t>()),
+    MatMulInteger<int8_t, int8_t>);
+
+template <>
+Status MatMulInteger<int8_t, int8_t>::PadMatrix(
+    int64_t row,
+    int64_t col,
+    int64_t align_size,
+    const int8_t*& src,
+    int64_t& pad_size,
+    IAllocatorUniquePtr<int8_t>& temp_mem_holder) const {
+  pad_size = align_size - col % align_size;
+  if (pad_size != align_size) {
+    temp_mem_holder = GetScratchBuffer<int8_t>(row * (col + pad_size));
+    ORT_RETURN_IF_ERROR(PadMatrixInLeadingDimension(src, temp_mem_holder.get(), row, col, pad_size));
+    src = temp_mem_holder.get();
+  } else {
+    pad_size = 0;
+  }
+
+  return Status::OK();
+}
+
+template <>
+Status MatMulInteger<int8_t, int8_t>::ComputeInternal(OpKernelContext* ctx) const {
+  auto a = ctx->Input<Tensor>(0);
+  auto b = ctx->Input<Tensor>(1);
+  ORT_ENFORCE(a != nullptr && b != nullptr);
+
+  MatMulComputeHelper helper;
+  ORT_RETURN_IF_ERROR(helper.Compute(a->Shape(), b->Shape()));
+  Tensor* Y = ctx->Output(0, helper.OutputShape());
+
+  const int8_t* a_ptr = a->template Data<int8_t>();
+  const int8_t* b_ptr = b->template Data<int8_t>();
+  int32_t* output_ptr = Y->template MutableData<int32_t>();
+
+  // validate zero points
+  const int8_t* a_offset = nullptr;
+  const int8_t* b_offset = nullptr;
+  if (has_a_zero_point_) {
+    auto a_zero_point = ctx->Input<Tensor>(2);
+    ORT_ENFORCE(IsScalarOr1ElementVector(a_zero_point),
+                "MatmulInteger : input1 zero point must be a scalar or 1D tensor of size 1");
+    a_offset = a_zero_point->template Data<int8_t>();
+  }
+  if (has_b_zero_point_) {
+    auto b_zero_point = ctx->Input<Tensor>(3);
+    ORT_ENFORCE(IsScalarOr1ElementVector(b_zero_point),
+                "MatmulInteger : input2 zero point must be a scalar or 1D tensor of size 1");
+    b_offset = b_zero_point->template Data<int8_t>();
+  }
```

Comment:
Slightly confused by the spec as it describes a_zero_point as a 'zero point tensor' and b_zero_point as a 'scale tensor'. 

>b_zero_point (optional) : T2
>Scale tensor for input 'B'. 

Are they both used as a zero point tensors here?

Reply:
> Slightly confused by the spec as it describes a_zero_point as a 'zero point tensor' and b_zero_point as a 'scale tensor'.
> 
> > b_zero_point (optional) : T2
> > Scale tensor for input 'B'.
> 
> Are they both used as a zero point tensors here?

a_zero_point and b_zero_point is the zero point for Matrix A and B respectively, and a_scale and b_scale is the scale for A and B.